lib/Parse/Lexer.cpp

```diff
@@ -1163,9 +1176,13 @@ unsigned Lexer::lexCharacter(const char *&CurPtr, char StopQuote,
     }
     // Move the pointer back to EOF.
     --CurPtr;
-    LLVM_FALLTHROUGH;
+    if (EmitDiagnostics)
+      diagnose(CurPtr-1, diag::lex_unterminated_string);
+    return ~1U;
   case '\n':  // String literals cannot have \n or \r in them.
   case '\r':
+    if (Modifiers & StringLiteralMultiline) // ... unless they are multiline
+      return CurPtr[-1];
```

Comment:
The alternative is some very subtle code on line 1711 which would also need to be replicated in the parsing phase lexStringLiteral() for error reporting.